### PR TITLE
Screen time storyteller pipe

### PR DESCRIPTION
**Pull request by onyedikachi-david**
# [PR] Implement Screen Time Storyteller Pipe

## Description
This PR implements a new pipe called "Screen Time Storyteller" for Screenpipe. The pipe analyzes user's 24/7 screen recordings and generates personalized, humorous roasts about their digital habits. It aims to provide insights and motivation for improving productivity through AI-powered feedback.

Related issue: #365 
Fixes: #365
/claim #365 

## Type of change
- [x] New feature (non-breaking change which adds functionality)

## How to test
1. Checkout this branch
2. Install dependencies
3. Run Screenpipe with the new Screen Time Storyteller pipe enabled
4. Check the generated narrative summary for personalized insights and humor

## Screenshot
Screenshot 2024-10-08 at 00 35 25

## Additional notes
A YouTube video demonstrating the Screen Time Storyteller pipe has been created and published at [https://youtu.be/XMVvkk0M7jQ](https://youtu.be/XMVvkk0M7jQ). The video showcases the functionality and DIY

No Reddit post was made; not a frequent Redditor so i don't have enough karma to post.

### CODE REVIEW

File: examples/typescript/pipe-screen-time-storyteller/pipe.ts [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+    content: string;
+    timestamp: string;
+    // Add other relevant fields
+}
+
+async function getAIProvider(config: Config): Promise<AIProvider> {
+    return { provider: config.aiProvider, model: config.models[config.aiProvider] };
+}
+
+async function queryScreenpipe(params: ScreenpipeQueryParams): Promise<ScreenpipeResponse | null> {
```
Comment (louis030195) [inline examples/typescript/pipe-screen-time-storyteller/pipe.ts:47]: you don't need this

use [pipe.queryScreenpipe](https://github.com/mediar-ai/screenpipe/blob/e4d26ed9ea59328f092705f5b7ddb5b2ce49a936/screenpipe-core/src/deno/runtime.js#L169)

**Comment by louis030195** [comment]
nice work!

let me time to test out

File: examples/typescript/pipe-screen-time-storyteller/pipe.ts (hunk 1 above)
Comment (onyedikachi-david) [reply examples/typescript/pipe-screen-time-storyteller/pipe.ts:47]: Okay.

File: examples/typescript/pipe-screen-time-storyteller/pipe.json [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  "description": "Generates a narrative summary of your day based on your screen activities",
+  "author": "AI Assistant",
+  "fields": [
+    {
+      "name": "aiProvider",
+      "value": "ollama"
+    },
+    {
+      "name": "models",
+      "value": {
```
Comment (louis030195) [inline examples/typescript/pipe-screen-time-storyteller/pipe.json:13]: does this work? did you test loading this pipe in the app UI?

it is not supposed to work like this


check example: https://github.com/mediar-ai/screenpipe/blob/main/examples/typescript/pipe-meeting-summary-by-email/pipe.json

default should be present, not value, and it will show as placeholder in app ui inputs

also object type does not exist, not sure how you tested this

File: examples/typescript/pipe-screen-time-storyteller/pipe.json [hunk 3]
```diff
@@ -0,0 +1,41 @@
+{
+  "name": "Screen Time Storyteller",
+  "version": "1.0.0",
+  "description": "Generates a narrative summary of your day based on your screen activities",
+  "author": "AI Assistant",
+  "fields": [
+    {
+      "name": "aiProvider",
+      "value": "ollama"
```
Comment (louis030195) [inline examples/typescript/pipe-screen-time-storyteller/pipe.json:9]: also type missing

File: examples/typescript/pipe-screen-time-storyteller/pipe.ts [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+                },
+                body: JSON.stringify({
+                    model: provider.model,
+                    messages: [{ role: "user", content: prompt }],
+                    max_tokens: 1000,
+                }),
+            });
+            result = await response.json();
+            console.log("Claude API response:", JSON.stringify(result, null, 2));
+            if (result.content && Array.isArray(result.content) && result.content[0] && result.content[0].text) {
```
Comment (louis030195) [inline examples/typescript/pipe-screen-time-storyteller/pipe.ts:101]: tip: avoid nesting if and try catch like this, it make the code very hard to read/change/understand

instead using early exit (or ask chatgpt how)

File: examples/typescript/pipe-screen-time-storyteller/pipe.json [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+      "description": "Type of content to process (ocr, audio, or all)"
+    },
+    {
+      "name": "githubToken",
+      "type": "string",
+      "default": "",
+      "description": "GitHub personal access token"
+    }
+  ],
+  "enabled": true
```
Comment (louis030195) [inline examples/typescript/pipe-screen-time-storyteller/pipe.json:62]: remove enabled

File: examples/typescript/pipe-screen-time-storyteller/README.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+1. Ensure you have Screenpipe installed and running.
+
+2. Copy the `screen-time-storyteller` folder into your Screenpipe pipes directory:
+   ```
+   cp -r screen-time-storyteller /path/to/your/screenpipe/pipes/
+   ```
+
+3. Edit the `pipe.json` file in the `screen-time-storyteller` folder with your API keys, GitHub token, and preferences:
+   ```json
```
Comment (louis030195) [inline examples/typescript/pipe-screen-time-storyteller/README.md:32]: outdated json

**Comment by cparish312** [comment]
@louis030195 @onyedikachi-david From my understanding this is just going to use the first 50 ocr results from the previous day. If that is the case it could be helpful to build a random or uniform option that grabs a random or uniformly distributed 100 ocr results or so within a time period. This could simplify pipes that want to do some sort of analysis over larger time periods.

**Comment by onyedikachi-david** [comment]
> @louis030195 @onyedikachi-david From my understanding this is just going to use the first 50 ocr results from the previous day. If that is the case it could be helpful to build a random or uniform option that grabs a random or uniformly distributed 100 ocr results or so within a time period. This could simplify pipes that want to do some sort of analysis over larger time periods.

Nice, that's a good suggestion. The main constraint is the AI context window, so even if it is a uniform time-based period, we're still limited in how many OCR results we can include. However, implementing a random or uniform sampling option could indeed provide better coverage over larger time periods.
I would have done more testing, but due to the cost of each iteration @cparish312

**Comment by onyedikachi-david** [comment]
Done @louis030195

**Comment by onyedikachi-david** [comment]
I am thinking of implementing another pipe that logs your activity on github readme daily.

**Comment by louis030195** [comment]
@onyedikachi-david for the cost DM me on discord i'll give you an openai key "louis030195"

**Comment by louis030195** [comment]
@onyedikachi-david hey yesterday did large refactor of how pipes work internally, which slightly change the API

like

pipe.queryScreenpipe -> queryScreenpipe

check pipes in the examples folder

basically we use Deno 

you can use dependencies now 

check this one to use dependencies https://github.com/mediar-ai/screenpipe/blob/main/examples/typescript/pipe-llama32-comment-linear-while-you-work/pipe.ts

this is how you use imports in deno https://github.com/mediar-ai/screenpipe/blob/main/examples/typescript/pipe-llama32-comment-linear-while-you-work/deno.json

rest did not change much, same way to install pipes etc 


also if you could use https://github.com/vercel/ai for the AI calls that would be awesome, it's an amazing lib and want people to use it to simplify dev 



hope it's not too much ask, then will merge and can move on to other pipes

**Comment by louis030195** [comment]
> @louis030195 @onyedikachi-david From my understanding this is just going to use the first 50 ocr results from the previous day. If that is the case it could be helpful to build a random or uniform option that grabs a random or uniformly distributed 100 ocr results or so within a time period. This could simplify pipes that want to do some sort of analysis over larger time periods.

good idea!

indeed we are still trying to find ways to fit days/weeks of data into LLM prompt which is hard

**Comment by onyedikachi-david** [comment]
> > @louis030195 @onyedikachi-david From my understanding this is just going to use the first 50 ocr results from the previous day. If that is the case it could be helpful to build a random or uniform option that grabs a random or uniformly distributed 100 ocr results or so within a time period. This could simplify pipes that want to do some sort of analysis over larger time periods.
> 
> good idea!
> 
> indeed we are still trying to find ways to fit days/weeks of data into LLM prompt which is hard

Use vector db search maybe?

**Comment by cparish312** [comment]
For a generic task like this pipeline or I'm working on building out a "here are 25 related articles" using exa.ai search not sure a vector db search is the best option. I'm working on some efficient larger scale summarization / topic extraction algos, such as de duplicating and clustering. Very open to any suggestions!

**Comment by louis030195** [comment]
> For a generic task like this pipeline or I'm working on building out a "here are 25 related articles" using exa.ai search, not sure a vector db search is the best option. I'm working on some efficient larger scale summarization / topic extraction algos, such as de duplicating and clustering. Very open to any suggestions!

we are working on adding embeddings & vector search in the future but the biggest issue is latency and resource usage, it should not overload people's computer and add delay for real time compute

we're going to experiment indexing recordings using embeddings at night or when the user is not using the computer, likely using multimodal 

vector search is still not perfect, it's just returning most similar result to the query, which does not solve all the search use cases but likely can improve some relevancy use cases 

something i experimented is a kind of recursive summarisation: https://github.com/mediar-ai/screenpipe/blob/main/examples/typescript/pipe-email-daily-log/pipe.ts

basically generate concise logs of tags, topics, summaries throughout the day and then feed all into the LLM and ask for a summary

i think this combined with vector search could be powerful

we could also experiment embeddings in the pipes, say embed a full day of data then generate a summary, something, before we have it in the core of screenpipe codebase

regarding deduplication i used previously either embeddings or string similarity heuristics, works well

**Comment by onyedikachi-david** [comment]
Screenshot 2024-10-12 at 19 16 05
Screenshot 2024-10-12 at 19 16 32

File: examples/typescript/pipe-screen-time-storyteller/screenpipe.d.ts [hunk 7]
```diff
@@ -0,0 +1,245 @@
+declare global {
```
Comment (louis030195) [inline examples/typescript/pipe-screen-time-storyteller/screenpipe.d.ts:1]: can you remove this file

**Comment by onyedikachi-david** [comment]
Done.

**Comment by louis030195** [comment]
@onyedikachi-david thanks  for the contribution :)

/approve
